src/payment_store.rs

```diff
@@ -0,0 +1,138 @@
+use crate::hex_utils;
+use crate::Error;
+
+use lightning::ln::{PaymentHash, PaymentPreimage, PaymentSecret};
+use lightning::util::persist::KVStorePersister;
+use lightning::{impl_writeable_tlv_based, impl_writeable_tlv_based_enum};
+
+use std::collections::HashMap;
+use std::iter::FromIterator;
+use std::ops::Deref;
+use std::sync::Mutex;
+
+/// Represents a payment.
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct PaymentInfo {
+	/// The payment hash, i.e., the hash of the `preimage`.
+	pub payment_hash: PaymentHash,
+	/// The pre-image used by the payment.
+	pub preimage: Option<PaymentPreimage>,
+	/// The secret used by the payment.
+	pub secret: Option<PaymentSecret>,
+	/// The amount transferred.
+	pub amount_msat: Option<u64>,
+	/// The direction of the payment.
+	pub direction: PaymentDirection,
+	/// The status of the payment.
+	pub status: PaymentStatus,
+}
+
+impl_writeable_tlv_based!(PaymentInfo, {
+	(0, payment_hash, required),
+	(1, preimage, required),
+	(2, secret, required),
+	(3, amount_msat, required),
+	(4, direction, required),
+	(5, status, required)
+});
+
+/// Represents the direction of a payment.
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum PaymentDirection {
+	/// The payment is inbound.
+	Inbound,
+	/// The payment is outbound.
+	Outbound,
+}
+
+impl_writeable_tlv_based_enum!(PaymentDirection,
+	(0, Inbound) => {},
+	(1, Outbound) => {};
+);
+
+/// Represents the current status of a payment.
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum PaymentStatus {
+	/// The payment is still pending.
+	Pending,
+	/// The payment suceeded.
+	Succeeded,
+	/// The payment failed.
+	Failed,
+}
+
+impl_writeable_tlv_based_enum!(PaymentStatus,
+	(0, Pending) => {},
+	(1, Succeeded) => {},
+	(2, Failed) => {};
+);
+
+/// The payment information will be persisted under this prefix.
+pub(crate) const PAYMENT_INFO_PERSISTENCE_PREFIX: &str = "payments";
+
+pub(crate) struct PaymentInfoStorage<K: Deref>
+where
+	K::Target: KVStorePersister,
+{
+	payments: Mutex<HashMap<PaymentHash, PaymentInfo>>,
+	persister: K,
+}
+
+impl<K: Deref> PaymentInfoStorage<K>
+where
+	K::Target: KVStorePersister,
+{
+	pub(crate) fn from_payments(mut payments: Vec<PaymentInfo>, persister: K) -> Self {
+		let payments = Mutex::new(HashMap::from_iter(
+			payments.drain(..).map(|payment_info| (payment_info.payment_hash, payment_info)),
```

Comment:
Can use `into_iter()` instead of `drain(..)`, I believe, and remove `mut` from `payments`.